[VL] Gluten fails to satisfy the ordering required by dynamic partition column

### Backend

VL (Velox)

### Bug description

#### Example
```sql
set spark.sql.maxConcurrentOutputFileWriters = 0;
set spark.sql.sources.partitionOverwriteMode = DYNAMIC;

-- src table
DROP TABLE IF EXISTS sql_migration.tmp_gluten_repro_src;
CREATE TABLE sql_migration.tmp_gluten_repro_src (k string, v string, day string) STORED AS ORC;

INSERT INTO sql_migration.tmp_gluten_repro_src VALUES
  ('k1','v01','2026-06-01'), ('k2','v02','2026-06-02'),
  ('k3','v03','2026-06-01'), ('k4','v04','2026-06-02'),
  ('k5','v05','2026-06-01'), ('k6','v06','2026-06-02'),
  ('k7','v07','2026-06-01'), ('k8','v08','2026-06-02'),
  ('k9','v09','2026-06-01'), ('k10','v10','2026-06-02');

-- target table
DROP TABLE IF EXISTS sql_migration.tmp_gluten_repro_tgt;
CREATE TABLE sql_migration.tmp_gluten_repro_tgt (k string, m string)
  PARTITIONED BY (day string) STORED AS ORC;

-- insert into partitions dynamically
INSERT OVERWRITE TABLE sql_migration.tmp_gluten_repro_tgt PARTITION(day)
SELECT k, max(v) AS m, day
FROM sql_migration.tmp_gluten_repro_src
GROUP BY day, k;

```

```
Exception:
Caused by: org.apache.hadoop.fs.FileAlreadyExistsException: /xx/xx/sql_migration/tmp_gluten_repro_tgt/.spark-staging-46b3ca99-cab6-4c7e-9e12-dd6a8ad81bc9/_temporary/0/_temporary/attempt_202607081201073995302620617304707_0039_m_000000_2127/day=2026-06-02/part-00000-46b3ca99-cab6-4c7e-9e12-dd6a8ad81bc9.c000.snappy.orc for client 10.69.52.150 already exists
	at org.apache.hadoop.hdfs.server.namenode.FSDirWriteFileOp.startFile(FSDirWriteFileOp.java:386)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.startFileInt(FSNamesystem.java:2464)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.startFile(FSNamesystem.java:2361)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.create(NameNodeRpcServer.java:774)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.create(ClientNamenodeProtocolServerSideTranslatorPB.java:462)
	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:522)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1025)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:931)
	at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:870)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1730)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2763)

	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at org.apache.hadoop.ipc.RemoteException.instantiateException(RemoteException.java:121)
	at org.apache.hadoop.ipc.RemoteException.unwrapRemoteException(RemoteException.java:88)
	at org.apache.hadoop.hdfs.DFSOutputStream.newStreamForCreate(DFSOutputStream.java:284)
	at org.apache.hadoop.hdfs.DFSClient.create(DFSClient.java:1219)
	at org.apache.hadoop.hdfs.DFSClient.create(DFSClient.java:1198)
	at org.apache.hadoop.hdfs.DFSClient.create(DFSClient.java:1136)
	at org.apache.hadoop.hdfs.DistributedFileSystem$8.doCall(DistributedFileSystem.java:531)
	at org.apache.hadoop.hdfs.DistributedFileSystem$8.doCall(DistributedFileSystem.java:528)
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
	at org.apache.hadoop.hdfs.DistributedFileSystem.create(DistributedFileSystem.java:542)
	at org.apache.hadoop.hdfs.DistributedFileSystem.create(DistributedFileSystem.java:469)
	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:1123)
	at org.apache.hadoop.fs.FileSystem.create(FileSystem.java:1103)
	at org.apache.orc.impl.PhysicalFsWriter.<init>(PhysicalFsWriter.java:94)
	at org.apache.orc.impl.WriterImpl.<init>(WriterImpl.java:230)
	at org.apache.orc.OrcFile.createWriter(OrcFile.java:1040)
	at org.apache.spark.sql.execution.datasources.orc.OrcOutputWriter.<init>(OrcOutputWriter.scala:50)
	at org.apache.spark.sql.execution.datasources.orc.OrcFileFormat$$anon$1.newInstance(OrcFileFormat.scala:92)
	at org.apache.spark.sql.execution.datasources.BaseDynamicPartitionDataWriter.renewCurrentWriter(FileFormatDataWriter.scala:298)
	at org.apache.spark.sql.execution.datasources.DynamicPartitionDataSingleWriter.write(FileFormatDataWriter.scala:365)
	at org.apache.spark.sql.execution.datasources.FileFormatDataWriter.writeWithMetrics(FileFormatDataWriter.scala:85)
	at org.apache.spark.sql.execution.datasources.FileFormatDataWriter.writeWithIterator(FileFormatDataWriter.scala:92)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.$anonfun$executeTask$2(FileFormatWriter.scala:481)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1398)
	at org.apache.spark.sql.execution.datasources.FileFormatWriter$.executeTask(FileFormatWriter.scala:488)
	... 17 more
```
#### Advanced Description
```text
Gluten plan:
Execute InsertIntoHadoopFsRelationCommand (16)
   +- WriteFiles (15)
      +- VeloxColumnarToRow (14)
         +- ^ ProjectExecTransformer (12)
            +- ^ RegularHashAggregateExecTransformer (11)
               +- ^ InputIteratorTransformer (10)
                  +- AQEShuffleRead (8)
                     +- ShuffleQueryStage (7), Statistics(sizeInBytes=271.0 B, rowCount=10)
                        +- ColumnarExchange (6)
                           +- VeloxResizeBatches (5)
                              +- ^ ProjectExecTransformer (3)
                                 +- ^ FlushableHashAggregateExecTransformer (2)
                                    +- ^ FileSourceScanExecTransformer orc spark_catalog.sql_migration.tmp_gluten_repro_src (1)

initial plan:
   Execute InsertIntoHadoopFsRelationCommand (24)
   +- WriteFiles (23)
      +- SortAggregate (22)
         +- Sort (21)
            +- Exchange (20)
               +- SortAggregate (19)
                  +- Sort (18)
                     +- Scan orc spark_catalog.sql_migration.tmp_gluten_repro_src (17)
```
Gluten transforms SortAggregate into HashAggregate，it destroys the ordering requirement for dynamic partitions writing. When dynamicPartitionDataSingleWriter is adopted, Task may process data that is not sorted by the partition column, which leading the  FileAlreadyExistsException. Actually, when `spark.sql.maxConcurrentOutputFileWriters`  is greater than  the num of dynamic partitions, dynamicPartitionDataConcurrentWriter can be adopted, and the problem can be avoided. But, I do not think it is a graceful way.


### Gluten version

1.6.0

### Spark version

3.5.8

### Spark configurations

_No response_

### System information

_No response_

### Relevant logs

```bash

```

--- a/gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/EnsureLocalSortRequirements.scala
+++ b/gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/EnsureLocalSortRequirements.scala
@@ -17,10 +17,12 @@
 package org.apache.gluten.extension.columnar
 
 import org.apache.gluten.extension.columnar.heuristic.HeuristicTransform
+import org.apache.gluten.sql.shims.SparkShimLoader
 
 import org.apache.spark.sql.catalyst.expressions.SortOrder
 import org.apache.spark.sql.catalyst.rules.Rule
-import org.apache.spark.sql.execution.{SortExec, SparkPlan}
+import org.apache.spark.sql.execution.{ColumnarWriteFilesExec, SortExec, SparkPlan}
+import org.apache.spark.sql.execution.datasources.WriteFilesExec
 
 /**
  * This rule is similar with `EnsureRequirements` but only handle local `SortExec`.
@@ -33,6 +35,24 @@ import org.apache.spark.sql.execution.{SortExec, SparkPlan}
 object EnsureLocalSortRequirements extends Rule[SparkPlan] {
   private lazy val transform: HeuristicTransform = HeuristicTransform.static()
 
+  private def requiredChildOrdering(plan: SparkPlan): Seq[Seq[SortOrder]] = {
+    plan match {
+      // V1Writes assumes that the logical ordering it prepared is preserved in the physical plan,
+      // so WriteFilesExec does not expose requiredChildOrdering itself. Gluten may invalidate that
+      // ordering when it replaces a SortAggregateExec with a hash aggregate.
+      case writeFiles: WriteFilesExec
+          if ColumnarWriteFilesExec.OnNoopLeafPath.unapply(writeFiles).isEmpty =>
+        Seq(
+          SparkShimLoader.getSparkShims.getV1WriteRequiredOrdering(
+            writeFiles.child.output,
+            writeFiles.partitionColumns,
+            writeFiles.bucketSpec,
+            writeFiles.options,
+            writeFiles.staticPartitions.size))
+      case _ => plan.requiredChildOrdering
+    }
+  }
+
   private def addLocalSort(
       originalChild: SparkPlan,
       requiredOrdering: Seq[SortOrder]): SparkPlan = {
@@ -44,7 +64,7 @@ object EnsureLocalSortRequirements extends Rule[SparkPlan] {
   override def apply(plan: SparkPlan): SparkPlan = {
     plan.transformUp {
       case p =>
-        val newChildren = p.children.zip(p.requiredChildOrdering).map {
+        val newChildren = p.children.zip(requiredChildOrdering(p)).map {
           case (child, requiredOrdering) =>
             // If child.outputOrdering already satisfies the requiredOrdering,
             // we do not need to sort.

--- a/gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/EnsureLocalSortRequirements.scala
+++ b/gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/EnsureLocalSortRequirements.scala
@@ -42,8 +42,7 @@ object EnsureLocalSortRequirements extends Rule[SparkPlan] {
     val resolver = SQLConf.get.resolver
     val staticPartitionNames = writeFiles.staticPartitions.keys
     writeFiles.partitionColumns.takeWhile {
-      partitionColumn =>
-        staticPartitionNames.exists(resolver(_, partitionColumn.name))
+      partitionColumn => staticPartitionNames.exists(resolver(_, partitionColumn.name))
     }.size
   }
 

--- a/gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/EnsureLocalSortRequirements.scala
+++ b/gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/EnsureLocalSortRequirements.scala
@@ -23,6 +23,7 @@ import org.apache.spark.sql.catalyst.expressions.SortOrder
 import org.apache.spark.sql.catalyst.rules.Rule
 import org.apache.spark.sql.execution.{ColumnarWriteFilesExec, SortExec, SparkPlan}
 import org.apache.spark.sql.execution.datasources.WriteFilesExec
+import org.apache.spark.sql.internal.SQLConf
 
 /**
  * This rule is similar with `EnsureRequirements` but only handle local `SortExec`.
@@ -34,6 +35,17 @@ import org.apache.spark.sql.execution.datasources.WriteFilesExec
  */
 object EnsureLocalSortRequirements extends Rule[SparkPlan] {
   private lazy val transform: HeuristicTransform = HeuristicTransform.static()
+
+  private def numStaticPartitionCols(writeFiles: WriteFilesExec): Int = {
+    // HadoopFs writes include static partition columns in partitionColumns, while Hive writes may
+    // only include the partition columns that are present in the write query.
+    val resolver = SQLConf.get.resolver
+    val staticPartitionNames = writeFiles.staticPartitions.keys
+    writeFiles.partitionColumns.takeWhile {
+      partitionColumn =>
+        staticPartitionNames.exists(resolver(_, partitionColumn.name))
+    }.size
+  }
 
   private def requiredChildOrdering(plan: SparkPlan): Seq[Seq[SortOrder]] = {
     plan match {
@@ -48,7 +60,7 @@ object EnsureLocalSortRequirements extends Rule[SparkPlan] {
             writeFiles.partitionColumns,
             writeFiles.bucketSpec,
             writeFiles.options,
-            writeFiles.staticPartitions.size))
+            numStaticPartitionCols(writeFiles)))
       case _ => plan.requiredChildOrdering
     }
   }

--- a/gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/EnsureLocalSortRequirements.scala
+++ b/gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/EnsureLocalSortRequirements.scala
@@ -16,6 +16,7 @@
  */
 package org.apache.gluten.extension.columnar
 
+import org.apache.gluten.execution.GlutenPlan
 import org.apache.gluten.extension.columnar.heuristic.HeuristicTransform
 import org.apache.gluten.sql.shims.SparkShimLoader
 
@@ -65,11 +66,19 @@ object EnsureLocalSortRequirements extends Rule[SparkPlan] {
   }
 
   private def addLocalSort(
+      plan: SparkPlan,
       originalChild: SparkPlan,
       requiredOrdering: Seq[SortOrder]): SparkPlan = {
     // FIXME: HeuristicTransform is costly. Re-applying it may cause performance issues.
     val newChild = SortExec(requiredOrdering, global = false, child = originalChild)
-    transform.apply(newChild)
+    (plan, originalChild) match {
+      case (_, child: GlutenPlan) if child.supportsColumnar =>
+        transform.apply(newChild)
+      case (parent: GlutenPlan, _) if parent.supportsColumnar =>
+        transform.apply(newChild)
+      case _ =>
+        newChild
+    }
   }
 
   override def apply(plan: SparkPlan): SparkPlan = {
@@ -82,7 +91,7 @@ object EnsureLocalSortRequirements extends Rule[SparkPlan] {
             if (SortOrder.orderingSatisfies(child.outputOrdering, requiredOrdering)) {
               child
             } else {
-              addLocalSort(child, requiredOrdering)
+              addLocalSort(p, child, requiredOrdering)
             }
         }
         p.withNewChildren(newChildren)

--- a/gluten-ut/spark34/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenV1WriteCommandSuite.scala
+++ b/gluten-ut/spark34/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenV1WriteCommandSuite.scala
@@ -98,6 +98,41 @@ class GlutenV1WriteCommandSuite
   with GlutenV1WriteCommandSuiteBase
   with GlutenSQLTestsBaseTrait {
 
+  testGluten("GLUTEN-12474: preserve ordering for dynamic partition writes") {
+    withSQLConf(
+      "spark.sql.maxConcurrentOutputFileWriters" -> "0",
+      "spark.sql.sources.partitionOverwriteMode" -> "DYNAMIC") {
+      withTable("gluten_12474_src", "gluten_12474_tgt") {
+        sql(
+          """
+            |CREATE TABLE gluten_12474_src USING ORC AS
+            |SELECT concat('k', id) AS k, format_string('v%02d', id) AS v,
+            |  if(id % 2 = 1, '2026-06-01', '2026-06-02') AS day
+            |FROM range(0, 10)
+            |""".stripMargin)
+
+        sql(
+          """
+            |CREATE TABLE gluten_12474_tgt (k STRING, m STRING)
+            |USING ORC
+            |PARTITIONED BY (day STRING)
+            |""".stripMargin)
+
+        sql(
+          """
+            |INSERT OVERWRITE TABLE gluten_12474_tgt PARTITION (day)
+            |SELECT k, max(v) AS m, day
+            |FROM gluten_12474_src
+            |GROUP BY day, k
+            |""".stripMargin)
+
+        checkAnswer(
+          sql("SELECT k, m, day FROM gluten_12474_tgt"),
+          sql("SELECT k, v, day FROM gluten_12474_src"))
+      }
+    }
+  }
+
   testGluten(
     "SPARK-41914: v1 write with AQE and in-partition sorted - non-string partition column") {
     withSQLConf(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true") {

--- a/gluten-ut/spark35/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenV1WriteCommandSuite.scala
+++ b/gluten-ut/spark35/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenV1WriteCommandSuite.scala
@@ -100,6 +100,41 @@ class GlutenV1WriteCommandSuite
   with GlutenSQLTestsBaseTrait
   with GlutenColumnarWriteTestSupport {
 
+  testGluten("GLUTEN-12474: preserve ordering for dynamic partition writes") {
+    withSQLConf(
+      "spark.sql.maxConcurrentOutputFileWriters" -> "0",
+      "spark.sql.sources.partitionOverwriteMode" -> "DYNAMIC") {
+      withTable("gluten_12474_src", "gluten_12474_tgt") {
+        sql(
+          """
+            |CREATE TABLE gluten_12474_src USING ORC AS
+            |SELECT concat('k', id) AS k, format_string('v%02d', id) AS v,
+            |  if(id % 2 = 1, '2026-06-01', '2026-06-02') AS day
+            |FROM range(0, 10)
+            |""".stripMargin)
+
+        sql(
+          """
+            |CREATE TABLE gluten_12474_tgt (k STRING, m STRING)
+            |USING ORC
+            |PARTITIONED BY (day STRING)
+            |""".stripMargin)
+
+        sql(
+          """
+            |INSERT OVERWRITE TABLE gluten_12474_tgt PARTITION (day)
+            |SELECT k, max(v) AS m, day
+            |FROM gluten_12474_src
+            |GROUP BY day, k
+            |""".stripMargin)
+
+        checkAnswer(
+          sql("SELECT k, m, day FROM gluten_12474_tgt"),
+          sql("SELECT k, v, day FROM gluten_12474_src"))
+      }
+    }
+  }
+
   testGluten(
     "SPARK-41914: v1 write with AQE and in-partition sorted - non-string partition column") {
     withSQLConf(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "true") {

--- a/gluten-ut/spark40/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenV1WriteCommandSuite.scala
+++ b/gluten-ut/spark40/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenV1WriteCommandSuite.scala
@@ -100,6 +100,41 @@ class GlutenV1WriteCommandSuite
   with GlutenSQLTestsBaseTrait
   with GlutenColumnarWriteTestSupport {
 
+  testGluten("GLUTEN-12474: preserve ordering for dynamic partition writes") {
+    withSQLConf(
+      "spark.sql.maxConcurrentOutputFileWriters" -> "0",
+      "spark.sql.sources.partitionOverwriteMode" -> "DYNAMIC") {
+      withTable("gluten_12474_src", "gluten_12474_tgt") {
+        sql(
+          """
+            |CREATE TABLE gluten_12474_src USING ORC AS
+            |SELECT concat('k', id) AS k, format_string('v%02d', id) AS v,
+            |  if(id % 2 = 1, '2026-06-01', '2026-06-02') AS day
+            |FROM range(0, 10)
+            |""".stripMargin)
+
+        sql(
+          """
+            |CREATE TABLE gluten_12474_tgt (k STRING, m STRING)
+            |USING ORC
+            |PARTITIONED BY (day STRING)
+            |""".stripMargin)
+
+        sql(
+          """
+            |INSERT OVERWRITE TABLE gluten_12474_tgt PARTITION (day)
+            |SELECT k, max(v) AS m, day
+            |FROM gluten_12474_src
+            |GROUP BY day, k
+            |""".stripMargin)
+
+        checkAnswer(
+          sql("SELECT k, m, day FROM gluten_12474_tgt"),
+          sql("SELECT k, v, day FROM gluten_12474_src"))
+      }
+    }
+  }
+
   // TODO: fix in Spark-4.0
   ignoreGluten(
     "SPARK-41914: v1 write with AQE and in-partition sorted - non-string partition column") {

--- a/gluten-ut/spark41/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenV1WriteCommandSuite.scala
+++ b/gluten-ut/spark41/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenV1WriteCommandSuite.scala
@@ -100,6 +100,41 @@ class GlutenV1WriteCommandSuite
   with GlutenSQLTestsBaseTrait
   with GlutenColumnarWriteTestSupport {
 
+  testGluten("GLUTEN-12474: preserve ordering for dynamic partition writes") {
+    withSQLConf(
+      "spark.sql.maxConcurrentOutputFileWriters" -> "0",
+      "spark.sql.sources.partitionOverwriteMode" -> "DYNAMIC") {
+      withTable("gluten_12474_src", "gluten_12474_tgt") {
+        sql(
+          """
+            |CREATE TABLE gluten_12474_src USING ORC AS
+            |SELECT concat('k', id) AS k, format_string('v%02d', id) AS v,
+            |  if(id % 2 = 1, '2026-06-01', '2026-06-02') AS day
+            |FROM range(0, 10)
+            |""".stripMargin)
+
+        sql(
+          """
+            |CREATE TABLE gluten_12474_tgt (k STRING, m STRING)
+            |USING ORC
+            |PARTITIONED BY (day STRING)
+            |""".stripMargin)
+
+        sql(
+          """
+            |INSERT OVERWRITE TABLE gluten_12474_tgt PARTITION (day)
+            |SELECT k, max(v) AS m, day
+            |FROM gluten_12474_src
+            |GROUP BY day, k
+            |""".stripMargin)
+
+        checkAnswer(
+          sql("SELECT k, m, day FROM gluten_12474_tgt"),
+          sql("SELECT k, v, day FROM gluten_12474_src"))
+      }
+    }
+  }
+
   // TODO: fix in Spark-4.0
   ignoreGluten(
     "SPARK-41914: v1 write with AQE and in-partition sorted - non-string partition column") {

--- a/shims/common/src/main/scala/org/apache/gluten/sql/shims/SparkShims.scala
+++ b/shims/common/src/main/scala/org/apache/gluten/sql/shims/SparkShims.scala
@@ -24,7 +24,8 @@ import org.apache.spark.broadcast.Broadcast
 import org.apache.spark.internal.io.FileCommitProtocol
 import org.apache.spark.sql.{AnalysisException, SparkSession}
 import org.apache.spark.sql.catalyst.InternalRow
-import org.apache.spark.sql.catalyst.expressions.{Attribute, BinaryArithmetic, Expression, InputFileBlockLength, InputFileBlockStart, InputFileName, RaiseError, UnBase64}
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
+import org.apache.spark.sql.catalyst.expressions.{Attribute, BinaryArithmetic, Expression, InputFileBlockLength, InputFileBlockStart, InputFileName, RaiseError, SortOrder, UnBase64}
 import org.apache.spark.sql.catalyst.plans.JoinType
 import org.apache.spark.sql.catalyst.plans.QueryPlan
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
@@ -128,6 +129,16 @@ trait SparkShims {
   }
 
   def enableNativeWriteFilesByDefault(): Boolean = false
+
+  // Planned V1 writes were introduced in Spark 3.4. Older versions do not expose a required
+  // ordering utility and keep the default empty ordering.
+  // TODO: Remove this shim after dropping Spark 3.3 support.
+  def getV1WriteRequiredOrdering(
+      outputColumns: Seq[Attribute],
+      partitionColumns: Seq[Attribute],
+      bucketSpec: Option[BucketSpec],
+      options: Map[String, String],
+      numStaticPartitionCols: Int): Seq[SortOrder] = Seq.empty
 
   def broadcastInternal[T: ClassTag](sc: SparkContext, value: T): Broadcast[T] = {
     // Since Spark 3.4, the `sc.broadcast` has been optimized to use `sc.broadcastInternal`.

--- a/shims/spark34/src/main/scala/org/apache/gluten/sql/shims/spark34/Spark34Shims.scala
+++ b/shims/spark34/src/main/scala/org/apache/gluten/sql/shims/spark34/Spark34Shims.scala
@@ -27,6 +27,7 @@ import org.apache.spark.paths.SparkPath
 import org.apache.spark.sql.{AnalysisException, SparkSession}
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.analysis.DecimalPrecision
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.aggregate._
 import org.apache.spark.sql.catalyst.plans.QueryPlan
@@ -207,6 +208,20 @@ class Spark34Shims extends SparkShims {
   }
 
   override def enableNativeWriteFilesByDefault(): Boolean = true
+
+  override def getV1WriteRequiredOrdering(
+      outputColumns: Seq[Attribute],
+      partitionColumns: Seq[Attribute],
+      bucketSpec: Option[BucketSpec],
+      options: Map[String, String],
+      numStaticPartitionCols: Int): Seq[SortOrder] = {
+    V1WritesUtils.getSortOrder(
+      outputColumns,
+      partitionColumns,
+      bucketSpec,
+      options,
+      numStaticPartitionCols)
+  }
 
   override def broadcastInternal[T: ClassTag](sc: SparkContext, value: T): Broadcast[T] = {
     SparkContextUtils.broadcastInternal(sc, value)

--- a/shims/spark35/src/main/scala/org/apache/gluten/sql/shims/spark35/Spark35Shims.scala
+++ b/shims/spark35/src/main/scala/org/apache/gluten/sql/shims/spark35/Spark35Shims.scala
@@ -27,6 +27,7 @@ import org.apache.spark.paths.SparkPath
 import org.apache.spark.sql.{AnalysisException, SparkSession}
 import org.apache.spark.sql.catalyst.{ExtendedAnalysisException, InternalRow}
 import org.apache.spark.sql.catalyst.analysis.DecimalPrecision
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.aggregate._
 import org.apache.spark.sql.catalyst.plans.QueryPlan
@@ -248,6 +249,20 @@ class Spark35Shims extends SparkShims {
   }
 
   override def enableNativeWriteFilesByDefault(): Boolean = true
+
+  override def getV1WriteRequiredOrdering(
+      outputColumns: Seq[Attribute],
+      partitionColumns: Seq[Attribute],
+      bucketSpec: Option[BucketSpec],
+      options: Map[String, String],
+      numStaticPartitionCols: Int): Seq[SortOrder] = {
+    V1WritesUtils.getSortOrder(
+      outputColumns,
+      partitionColumns,
+      bucketSpec,
+      options,
+      numStaticPartitionCols)
+  }
 
   override def broadcastInternal[T: ClassTag](sc: SparkContext, value: T): Broadcast[T] = {
     SparkContextUtils.broadcastInternal(sc, value)

--- a/shims/spark40/src/main/scala/org/apache/gluten/sql/shims/spark40/Spark40Shims.scala
+++ b/shims/spark40/src/main/scala/org/apache/gluten/sql/shims/spark40/Spark40Shims.scala
@@ -27,6 +27,7 @@ import org.apache.spark.paths.SparkPath
 import org.apache.spark.sql.{AnalysisException, SparkSession}
 import org.apache.spark.sql.catalyst.{ExtendedAnalysisException, InternalRow}
 import org.apache.spark.sql.catalyst.analysis.DecimalPrecisionTypeCoercion
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.aggregate._
 import org.apache.spark.sql.catalyst.plans.{JoinType, LeftSingle}
@@ -253,6 +254,20 @@ class Spark40Shims extends SparkShims {
   }
 
   override def enableNativeWriteFilesByDefault(): Boolean = true
+
+  override def getV1WriteRequiredOrdering(
+      outputColumns: Seq[Attribute],
+      partitionColumns: Seq[Attribute],
+      bucketSpec: Option[BucketSpec],
+      options: Map[String, String],
+      numStaticPartitionCols: Int): Seq[SortOrder] = {
+    V1WritesUtils.getSortOrder(
+      outputColumns,
+      partitionColumns,
+      bucketSpec,
+      options,
+      numStaticPartitionCols)
+  }
 
   override def broadcastInternal[T: ClassTag](sc: SparkContext, value: T): Broadcast[T] = {
     SparkContextUtils.broadcastInternal(sc, value)

--- a/shims/spark41/src/main/scala/org/apache/gluten/sql/shims/spark41/Spark41Shims.scala
+++ b/shims/spark41/src/main/scala/org/apache/gluten/sql/shims/spark41/Spark41Shims.scala
@@ -27,6 +27,7 @@ import org.apache.spark.paths.SparkPath
 import org.apache.spark.sql.{AnalysisException, SparkSession}
 import org.apache.spark.sql.catalyst.{ExtendedAnalysisException, InternalRow}
 import org.apache.spark.sql.catalyst.analysis.DecimalPrecisionTypeCoercion
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.aggregate._
 import org.apache.spark.sql.catalyst.plans.{JoinType, LeftSingle}
@@ -252,6 +253,20 @@ class Spark41Shims extends SparkShims {
   }
 
   override def enableNativeWriteFilesByDefault(): Boolean = true
+
+  override def getV1WriteRequiredOrdering(
+      outputColumns: Seq[Attribute],
+      partitionColumns: Seq[Attribute],
+      bucketSpec: Option[BucketSpec],
+      options: Map[String, String],
+      numStaticPartitionCols: Int): Seq[SortOrder] = {
+    V1WritesUtils.getSortOrder(
+      outputColumns,
+      partitionColumns,
+      bucketSpec,
+      options,
+      numStaticPartitionCols)
+  }
 
   override def broadcastInternal[T: ClassTag](sc: SparkContext, value: T): Broadcast[T] = {
     SparkContextUtils.broadcastInternal(sc, value)

